pkg/kv/kvserver/replica_application_result.go

```diff
@@ -505,6 +507,77 @@ func (r *Replica) stagePendingTruncationRaftMuLocked(pt pendingTruncation) {
 	r.asLogStorage().stagePendingTruncationRaftMuLocked(pt)
 }
 
+func (r *replicaLogStorage) stagePendingTruncationSharedRaftMuLockedMuLocked(pt pendingTruncation) {
+	// This code is shared between regular log truncations and snapshot-induced
+	// truncations.
+	r.shMu.trunc = pt.RaftTruncatedState
+	// Ensure the raft log size is not negative since it isn't persisted between
+	// server restarts.
+	// TODO(pav-kv): should we distrust the log size if it goes negative?
+	r.shMu.size = max(r.shMu.size+pt.logDeltaBytes, 0)
+	r.shMu.lastCheckSize = max(r.shMu.lastCheckSize+pt.logDeltaBytes, 0)
+	if !pt.isDeltaTrusted {
+		r.shMu.sizeTrusted = false
+	}
+}
+
+func (r *replicaLogStorage) stagePendingTruncationOnSnapshotRaftMuLocked(
+	truncState kvserverpb.RaftTruncatedState,
+) {
+	r.raftMu.AssertHeld()
+
+	// A snapshot application implies a log truncation to the snapshot's index,
+	// and we apply the resulting memory state here (before the snapshot takes
+	// effect, i.e. the log entries disappear from storage). This avoids
+	// situations in which entries were already removed, but the in-mem state
+	// indicates that they ought to still exist.
+	//
+	// The truncation finalized below, after the snapshot is visible.
+
+	// TODO(during review): when is it actually safe to clear the cache if we
```

Comment:
Hmm. So if we clear it here, some `mu` reader can cause it refilled before `r.mu.Lock()` in the next line happens. If we first `stage` and then, after releasing `mu`, clear the cache, we have a situation that entries in the cache are, worst case, in an inconsistent fork.

We are holding `raftMu` here, so no append attempts will happen in the meantime. Any racing reader will observe `RaftTruncatedState == LastEntryID`, and `Term()/Entries()` calls will be short circuited. There will be no access to storage or cache, so both can temporarily contain garbage.

Bottom line: I think it's safe to clear the cache after releasing `mu` but while still holding `raftMu`.